File: src/main/scala/ru/ifmo/model/Matrix.scala
package ru.ifmo.model

import java.util

class Matrix(m: Array[Array[Double]]) {
  val data: Array[Array[Double]] = m
  val rows: Int = m.length
  val cols: Int = m(0).length

  def mul(m: Matrix): Matrix = {
    val rows = this.rows
    val cols = m.cols
    val t = this.cols
    val result: Array[Array[Double]] = Array.ofDim(rows, cols)

    for (i <- 0 until rows; j <- 0 until cols) {
      var s = 0.0
      for (k <- 0 until t) {
        s += this.data(i)(k) * m.data(k)(j)
      }
      result(i)(j) = s
    }

    new Matrix(result)
  }

  def apply(i: Int, j: Int): Double = {
    data(i)(j)
  }

  override def toString: String = {
    data.foldLeft("")((str, array) => str + util.Arrays.toString(array) + '\n')
  }
}

File: src/main/java/ru/ifmo/docking/model/Matrix.java
package ru.ifmo.docking.model;

import java.util.Arrays;

public class Matrix {
    private final double[][] data;

    public int rowsCount() {
        return data.length;
    }

    public int columnsCount() {
        return data[0].length;
    }

    public Matrix(double[][] data) {
        this.data = data;
    }

    public double get(int i, int j) {
        return data[i][j];
    }

    public Matrix mul(Matrix that) {
        double[][] result = new double[this.rowsCount()][that.columnsCount()];
        for (int i = 0; i < this.rowsCount(); i++) {
            for (int j = 0; j < that.columnsCount(); j++) {
                for (int k = 0; k < this.columnsCount(); k++) {
                    result[i][j] += this.get(i, k) * that.get(k, j);
                }
            }
        }
        return new Matrix(result);
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder();
        for (double[] row : data) {
            sb.append(Arrays.toString(row)).append('\n');
        }
        return sb.toString();
    }
}
